pkg/sentry/fsimpl/erofs/directory.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package erofs
+
+import (
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/erofs"
+	"gvisor.dev/gvisor/pkg/errors/linuxerr"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+)
+
+func (d *dentry) isDir() bool {
+	return d.inode.IsDir()
+}
+
+func (d *dentry) getChild(ctx context.Context, name string) (*dentry, error) {
+	if !d.isDir() {
+		return nil, linuxerr.ENOTDIR
+	}
+
+	// Fast path.
+	d.dirMu.RLock()
+	child, ok := d.childMap[name]
```

Comment:
So we are duplicating the childMap in the dentry and the inode? This can cost more memory usage. `inode.childMap` seems to be an absolute map: if non-nil, contains all dirents. The `dentry.childMap` is populated on a need based basis as paths are walked on. I guess this is fine, gofer does something similar. But gofer only caches the **dirents** on IterDirents, not the entire inode.

Maybe we can do something like this:
- Let inode provide methods like:
  - Lookup: will do binary search on the underlying blocks, since dirents are sorted. It will construct an `*Inode` and return it. This will be called from `dentry.getChild()`.
  - IterDirents: will somewhat do what `ensureChildMap()` is doing right now. It will iterate through all dirents and call a callback function with the `erofs.Dirent` read from disk (**not** the inode). This will be called from `directoryFD.IterDirents()`.
- On the first call to `inode.IterDirents`, we can cache all the dirents in `inode.dirents`. So future calls to `IterDirents` and `Lookup` can just return results using this.

As of right now, a lookup for any file will cause **all** dirents to be read and cached and their corresponding Inodes to be also read from disk. This can cause performance issues on large directories. IterDirents is also currently doing more work than needed; i.e. reading each Inode from disk, in addition to the Dirent. Not super familiar with how inodes are packed on disk, but such premature reading of all Inodes can cause random access across the image, which can lead to unnecessary page faults.

Reply:
Oh I see that we are also caching `dirents` in `directoryFD`. We should get rid of that and instead cache dirents in the `inode`, as described above. This is also beneficial because the same directory inode can shared by many directory FDs, and we don't want a separate cache for all FDs.

Reply:
You are right. Currently, the caching policy looks strange and it bothers me as well.. I will rewrite this part in the next version.

Reply:
In linux kernel, we cache directory blocks in page cache, but it can be reclaimed due to memory pressure of course.
All accessed dirents (no matter positive or nagative) will be cached into vfs dcache as well on demand..

Reply:
> So maybe you could also cause Lookup to read all dirents and cache them (but this hurts large directories).

I will give it a try.

> But I guess the larger point I wanted to make was that we should not be constructing Inodes while caching dirents, because that can be expensive.

Yeah, I got that point, but forgot to respond it. It's a good point. Even if we cache the dirents, there is no need to construct the inodes, otherwise it will waste memory and may also trigger unnecessary page faults. Maybe we just need the name and nid (which allows us to locate the inode quickly) in this case, I will take a deeper look.

Besides, the whole image is mapped into sentry via a readonly/shared mapping, maybe we just need to let host kernel cache these blocks/pages transparently for us. It may work fine, unless the memory pressure is too high.